### PR TITLE
Fix found issues and keep legacy permission

### PR DESCRIPTION
**Pull request by alexcrea**
Fix of the issues I found. 
Still working on it.

### CODE REVIEW

**Comment by alexcrea** [comment]
Enchantment name being a color similar to lore may not be ideal.
image

**Comment by alexcrea** [comment]
Also. I do small commit for the ease of review. But It can be squashed and merged for a smaller commit history.

**Comment by alexcrea** [comment]
Caching enchantment list (last commit I pushed) provide clear performance gain:

Tested with a sword/enchanted book with the same 3 enchantment
every test has 10,000 call and running on my potato pc (spigot 1.21.1):

Baseline (only sword on the anvil. for sanity check)
Tested Disenchantment: 17 ms
Tested Shatterment: 13 ms

Tested Disenchantment (no cache)
Result: 309-345-398-314-422-422-361-656-348 ms (9 trys)
min-avg-max: 309-397-656 ms

Tested Shatterment (no cache)
Result: 996-504-305-288-719-431-327-361-611 ms (9 trys)
min-avg-max: 288-504-996 ms


Tested Disenchantment (with cache)
Result:  296-159-130-103-121-159-102-144-180 ms (9 trys)
min-avg-max: 102-154-296 ms
avg delta: +243 ms (~2.5 time faster)

Tested Shatterment (with cache)
Result: 165-150-121-104-103-136-92-106-121 ms (9 trys)
min-avg-max: 92-122-165 ms
avg delta: +382 ms (~4 time faster)

Result: even if my test is not ideal and have a lot of flaw, benefice of caching echantments is so big that it can't be interpreted as a test flaw.

**Comment by H7KZ** [comment]
Got quick look at the caching and how does it perform in terms of memory usage? The disenchantment plugin won't be used as often to make a cache in it, only if the server is huge, in that case there would be better to have a config option I think :DD

**Comment by H7KZ** [comment]
> Enchantment name being a color similar to lore may not be ideal. image

Yeah thats weird, I thought about having it similar across all the views but this feels off, if you put some color that looks good then amazing!

**Comment by H7KZ** [comment]
> Also. I do small commit for the ease of review. But It can be squashed and merged for a smaller commit history.

It alright, when I do come back from vacation I will review and try to merge soon as possible

**Comment by H7KZ** [comment]
Screenshot_20240826_064221_Chrome

Hate these type of people, cannot even create an issue and gives 1 star rating, mfs 💀 Why the fuck everyone thinks I do have the time to test every single thing on the plugin before push, I have my life also 💀💀😭😭😭 Well.... we droppin support for squared enchants for now xddd

**Comment by alexcrea** [comment]
> Got quick look at the caching and how does it perform in terms of memory usage? The disenchantment plugin won't be used as often to make a cache in it, only if the server is huge, in that case there would be better to have a config option I think :DD GREAT WORK SO FAR!!!

According to [this](https://developer.ibm.com/articles/j-codetoheap/#java-collections-hashmap8). 
Assuming 32bit JVM (as the article is written with that in mind). Assuming we store every vanilla enchantment (42) the size of the internal array would be 64. 32bit (node object) + 4 byte (node ref) time 42 = 1512 bytes for the node objects. + 128 bytes ("empty" hash map object size) + 64 bytes (Overhead) = 1704 bytes. (under 2kb) (I will not calculate for 64bit, but worst scenario would be 4kb. Probably closer to 3kb.)
And don't forget we have 2 maps, so about 4kb (probably about 6kb for 64bit JVM)
Please note, I may have made an error somewhere, as I do not certain I understand how java memory work. So may not be 100% accurate (reason why I rounded 1704 to 2kb and making it an upper limit).

Also, it is lazy. If the server never needed the map (since restart), it would have not been generated. 
I think it's okay to trade that amount of memory for a 2x performance gain on usage. (about 0.1% of a tick)

**Comment by alexcrea** [comment]
> Screenshot_20240826_064221_Chrome
> 
> Hate these type of people, cannot even create an issue and gives 1 star rating, mfs 💀 Why the fuck everyone thinks I do have the time to test every single thing on the plugin before push, I have my life also 💀💀😭😭😭 Well.... we droppin support for squared enchants for now xddd

Sad that they did not bother to open as issue.
Like, how do you want us to fix the issue you encounter if you do not even tell us there is one ?

But I like the developer of Enchantment² (and was the first custom enchantment plugin I supported). So I like to work on supporting it again later.

Also the plugins name is wrong on spigot (`Disenchantment [1.21.1 - 1.20.6]` should probably be `Disenchantment [1.21.1]`)

**Comment by alexcrea** [comment]
Made some important changes of how the plugin got the enchantment list from an item and globally.
This is so it can fetch a dummy Enchantment instance from plugin that do not register their enchantment in the registry.
For example. Enchantment Squared use item custom data to store their enchantments.

From in game test. It looks like everything work as expected.

File: src/main/java/cz/kominekjan/disenchantment/config/migrations/PermissionMigration.java
Comment (H7KZ) [inline src/main/java/cz/kominekjan/disenchantment/config/migrations/PermissionMigration.java:1]: Migrations are only for config, maybe it will be better to create its own class for permission management, where there will be all values, even legacy. Divided in classes like Current and Legacy, then methods of the events etc. to auto fill the required permissions idk

Comment (alexcrea) [reply src/main/java/cz/kominekjan/disenchantment/config/migrations/PermissionMigration.java:1]: We could do classes for legacy permission. But the goal of current Permission Migration class was to temporary support old permission and warn admin that it will be removed.
Not the best package for this class tbh.

Comment (alexcrea) [reply src/main/java/cz/kominekjan/disenchantment/config/migrations/PermissionMigration.java:1]: I started working on it. But my approach on the problem is a little different:
A permission goal enum where we could do `PermissionGoal.Disenchantment.haspermission(player)`
Permission goal enum would only store an array of PermissionPart And would check every of them until one is found correct or return false.
A PermissionNode contain a permission string (tested permission) & Permission type (enum. would contain DEFAULT & LEGACY but expendable)
PermissionNode with LEGACY PermissionType would print the "caution you are using legacy permission" warning.

This approach would make easy the need to had or modify permission, as it is just changing an enum.
Would also cover future need as we can just use new permission type

**Comment by alexcrea** [comment]
I added "disenchantment.gui.all" and "disenchantment.command.all".
Also, "disenchantment.gui" was used but not present in plugin.yml or PERMISSION.md

Tell me if you see any thing I may have missed. and if this system is okay for you

**Comment by H7KZ** [comment]
small thing I see you make spaces at the end of the files between two closing tags `} }` why? :D I would like to keep the codebase same and do not mix up formatting...

and if you do use the jetbrains intellij then there is a great tooling for somewhat making the code formatting a little bit similar "reformat code" 
image

**Comment by H7KZ** [comment]
image

File: src/main/java/cz/kominekjan/disenchantment/utils/DisenchantUtils.java [hunk 1]
```diff
@@ -50,4 +56,40 @@ public static Boolean canBeEnchanted(Material material) {
...
+            enchantments = new HashMap<>(item.getEnchantments());
+        }
+
+        HashMap<String, IPlugin> activatedPlugins = PluginManager.getActivatedPlugins();
+
+        if (!activatedPlugins.isEmpty()) {
+            for (IPlugin plugin : activatedPlugins.values()) {
+                plugin.fetchComplementaryEnchantment(item, enchantments);
+            }
+        }
```
Comment (H7KZ) [inline src/main/java/cz/kominekjan/disenchantment/utils/DisenchantUtils.java:75]: this here for future plugins or....? Rn does not do anything, default method fetchComplementaryEnchantment is not implemented on any of the supported plugins

File: src/main/java/cz/kominekjan/disenchantment/utils/EventCheckUtils.java [hunk 2]
```diff
@@ -5,30 +5,40 @@
...
-import org.bukkit.inventory.meta.EnchantmentStorageMeta;
 
+import java.util.Collections;
 import java.util.Map;
 
 public class EventCheckUtils {
     public static class Disenchantment {
-        public static Boolean isEventValid(ItemStack firstItem, ItemStack secondItem) {
-            if (firstItem == null || secondItem == null) return false;
+        public static Map<Enchantment, Integer> getValidEnchantments(ItemStack firstItem, ItemStack secondItem) {
```
Comment (H7KZ) [inline src/main/java/cz/kominekjan/disenchantment/utils/EventCheckUtils.java:14]: good idea with returning a map of the valid enchantments!

File: src/main/java/cz/kominekjan/disenchantment/plugins/IPlugin.java [hunk 3]
```diff
@@ -12,4 +14,13 @@ public interface IPlugin {
 
     ItemStack removeEnchantments(ItemStack firstItem, Map<Enchantment, Integer> enchantments);
 
+    default void activate() {
+    }
```
Comment (H7KZ) [inline src/main/java/cz/kominekjan/disenchantment/plugins/IPlugin.java:18]: why is this here when it is not used? I am confused

File: src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java [hunk 4]
```diff
@@ -107,22 +104,23 @@ public final GUIItem[] getEnchantmentsItems(List<Enchantment> enchantments) {
 
             if (enchantment == null) continue;
 
-            EnchantmentState disenchantmentState = disenchantmentEnchantmentsStates.getOrDefault(enchantment, EnchantmentState.ENABLED);
-            EnchantmentState shattermentState = shattermentEnchantmentsStates.getOrDefault(enchantment, EnchantmentState.ENABLED);
+            EnchantmentState[] disenchantmentState = new EnchantmentState[]{disenchantmentEnchantmentsStates.getOrDefault(enchantment, EnchantmentState.ENABLED)};
+            EnchantmentState[] shattermentState = new EnchantmentState[]{shattermentEnchantmentsStates.getOrDefault(enchantment, EnchantmentState.ENABLED)};
```
Comment (H7KZ) [inline src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java:108]: officially confused... why array?

File: src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java [hunk 5]
```diff
@@ -107,22 +104,23 @@ public final GUIItem[] getEnchantmentsItems(List<Enchantment> enchantments) {
...
 
             worldItems[i] = new GUIItem(
                     freeSlots[i],
                     DefaultGUIElements.enchantmentItem(
                             ChatColor.GRAY + "" + ChatColor.BOLD + enchantment.getKey().getKey(),
-                            ChatColor.GRAY + "Disenchantment: " + disenchantmentState.getDisplayName(),
-                            ChatColor.GRAY + "Shatterment: " + shattermentState.getDisplayName()
+                            ChatColor.GRAY + "Disenchantment: " + disenchantmentState[0].getDisplayName(),
+                            ChatColor.GRAY + "Shatterment: " + disenchantmentState[0].getDisplayName()
```
Comment (H7KZ) [inline src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java:115]: both are disenchantment? :DDD

File: src/main/java/cz/kominekjan/disenchantment/events/ShatterEvent.java [hunk 6]
```diff
@@ -80,10 +62,11 @@ public void onDisenchantmentEvent(PrepareAnvilEvent e) {
         HashMap<String, IPlugin> activatedPlugins = PluginManager.getActivatedPlugins();
 
         if (activatedPlugins.isEmpty()) {
-            book = VanillaPlugin.createEnchantedBook(enchantments);
+            book = VanillaPlugin.createEnchantedBook(randomEnchantmentShatter);
         } else {
             for (IPlugin plugin : activatedPlugins.values()) {
-                book = plugin.createEnchantedBook(enchantments);
+                // won't 2 plugin activated on the same time would only use the last book ?
```
Comment (H7KZ) [inline src/main/java/cz/kominekjan/disenchantment/events/ShatterEvent.java:68]: I dont think anyone would be brave enough to have more than 1 custom enchantment plugin running, cant imagine the amount of times it makes a collision or something :DDDD

File: src/main/java/cz/kominekjan/disenchantment/events/ShatterEvent.java [hunk 7]
```diff
@@ -17,44 +17,28 @@
...
-        EnchantmentStorageMeta firstItemItemMeta = (EnchantmentStorageMeta) firstItem.getItemMeta();
-
-        HashMap<Enchantment, Integer> enchantments = new HashMap<>(firstItemItemMeta.getStoredEnchants());
-
-        if (enchantments.size() < 2) return;
-
-        Config.Shatterment.getEnchantmentStates().forEach((enchantment, state) -> {
-            if (EnchantmentState.KEEP.equals(state)) enchantments.remove(enchantment);
-        });
+        Map<Enchantment, Integer> enchantments = EventCheckUtils.Shatterment.getValidEnchantments(firstItem, secondItem);
```
Comment (H7KZ) [inline src/main/java/cz/kominekjan/disenchantment/events/ShatterEvent.java:41]: nice!

File: src/main/java/cz/kominekjan/disenchantment/config/Config.java [hunk 8]
```diff
@@ -229,25 +250,42 @@ public static Boolean setDisabledWorlds(List<World> worlds) {
...
+                }
+
                 enchantmentStates.put(
-                        enchantments.stream().filter(e -> e.getKey().getKey().equalsIgnoreCase(enchantment)).findFirst().orElse(null),
+                        enchantment,
                         state
                 );
             }
 
+            STATE_CACHE = enchantmentStates;
```
Comment (H7KZ) [inline src/main/java/cz/kominekjan/disenchantment/config/Config.java:287]: alr, the state cache was a good idea, +rep

**Comment by alexcrea** [comment]
> small thing I see you make spaces at the end of the files between two closing tags `} }` why? :D I would like to keep the codebase same and do not mix up formatting...
> 
> and if you do use the jetbrains intellij then there is a great tooling for somewhat making the code formatting a little bit similar "reformat code" image

I'm still not used to intellij toolling. Will definitivly use that now ! thx

**Comment by alexcrea** [comment]
> image

Sadly it is used on one of the unused depecated function I had to implement to create a dummy enchantment object.
That mean there will be not call of function depending on that deprecated import. making it, probably, working for future version even if the class is deleted (At least I hope)

Only alternative I known is to rewrite most of the part of the plugin that use Enchantment and basicly use a kind of wrapped enchantment that either a registered enchantment or a enchantment squared enchantments (I did that for custom anvil for example)
But it would mean a lot of rewrite. so I think just checking what function was keeped and deleted on Enchantments when new paper minecraft version release is 

Futhermore if it create issue. it would be only for enchantment squared user updating there server before we find the issue and not everyone.

File: src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java (hunk 5 above)
Comment (alexcrea) [reply src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java:115]: oups

File: src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java (hunk 4 above)
Comment (alexcrea) [reply src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java:108]: We need to change the stored EnchantmentState in a lambda as we only care about previous state and not the state at the gui creation.
lambda need final variable. that mean we can't just store a disenchantmentState and change it as the variable would not be considered final.
So, we either need: 
- An array to store the current object (the array will be final but not what stored inside). current implementation
- an AtomicReference<EnchantmentState>.  previously implemented but you deleted it so I guessed that was not a good idea.
- Any other type of final reference to a reference

Comment (H7KZ) [reply src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java:108]: oh... well, in that case that atomic reference isn't bad :D

Comment (alexcrea) [reply src/main/java/cz/kominekjan/disenchantment/guis/impl/EnchantmentsGUI.java]: Done

**Comment by H7KZ** [comment]
Move to testing stage? :DDD Once I find time I will build it, refactor it through intelliJ tools, test it on server, how the config migration works and test out all commands & events with excellent enchants and squared enchants

this alright?

**Comment by alexcrea** [comment]
> Move to testing stage? :DDD Once I find time I will build it, refactor it through intelliJ tools, test it on server, how the config migration works and test out all commands & events with excellent enchants and squared enchants
> 
> this alright?

Look okay to me !
For my testing. everything look okay to me on spigot.

**Comment by H7KZ** [comment]
Tested it, squared enchantments do not work :( Could you have a look at that please? Thank you!

**Comment by alexcrea** [comment]
> Tested it, squared enchantments do not work :( Could you have a look at that please? Thank you!

Tested and it do work on my side. Using spigot. Will test on paper and edit this comment.

**Comment by H7KZ** [comment]
Amazing! So... everything tested I guess... we can merge and release? :D

**Comment by alexcrea** [comment]
I do think it is ready for release !
